scipy/optimize/_qap.py

```diff
@@ -0,0 +1,418 @@
+            P = (K + J) / 2
+        elif init_method == "barycenter":
+            P = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+
+        const_sum = A21 @ np.transpose(B21) + np.transpose(A12) @ B12
+        grad_P = math.inf  # gradient of P
+        n_iter = 0  # number of FW iterations
+
+        # OPTIMIZATION WHILE LOOP BEGINS
+        while grad_P > eps and n_iter < max_iter:
+            delta_f = (
+                const_sum + A22 @ P @ B22.T + A22.T @ P @ B22
+            )  # computing the gradient of f(P) = -tr(APB^tP^t)
+            rows, cols = linear_sum_assignment(
+                obj_func_scalar * delta_f
+            )  # run hungarian algorithm on gradient(f(P))
+            Q = np.zeros((n_unseed, n_unseed))
+            Q[rows, cols] = 1  # initialize search direction matrix Q
+
+            def f(x):  # computing the original optimization function
+                return obj_func_scalar * (
+                    np.trace(A11.T @ B11)
+                    + np.trace(np.transpose(x * P + (1 - x) * Q) @ A21 @ B21.T)
+                    + np.trace(np.transpose(x * P + (1 - x) * Q) @ A12.T @ B12)
+                    + np.trace(
+                        A22.T
+                        @ (x * P + (1 - x) * Q)
+                        @ B22
+                        @ np.transpose(x * P + (1 - x) * Q)
+                    )
+                )
+
+            alpha = minimize_scalar(
+                f, bounds=(0, 1), method="bounded"
+            ).x  # computing the step size
+            P_i1 = alpha * P + (1 - alpha) * Q  # Update P
+            grad_P = np.linalg.norm(P - P_i1)
+            P = P_i1
+            n_iter += 1
+        # end of FW optimization loop
+
+        row, col = linear_sum_assignment(
+            -P
+        )  # Project onto the set of permutation matrices
+        perm_inds_new = np.concatenate(
+            (np.arange(n_seeds), np.array([x + n_seeds for x in col]))
+        ).astype(int)
+
+        score_new = np.trace(
+            np.transpose(cost_matrix)
+            @ dist_matrix[np.ix_(perm_inds_new, perm_inds_new)]
+        )  # computing objective function value
+
+        if obj_func_scalar * score_new < obj_func_scalar * score:  # minimizing
+            score = score_new
+            perm_inds = np.zeros(n, dtype=int)
+            perm_inds[permutation_cost] = permutation_dist[perm_inds_new]
+
+    permutation_cost_unshuffle = _unshuffle(permutation_cost, n)
+    cost_matrix = cost_matrix[
+        np.ix_(permutation_cost_unshuffle, permutation_cost_unshuffle)
+    ]
+    permutation_dist_unshuffle = _unshuffle(permutation_dist, n)
+    dist_matrix = dist_matrix[
+        np.ix_(permutation_dist_unshuffle, permutation_dist_unshuffle)
+    ]
+
+    score = np.trace(
+        np.transpose(cost_matrix) @ dist_matrix[np.ix_(perm_inds, perm_inds)]
+    )
+
+    res = {"col_ind": perm_inds, "score": score, "nit": n_iter}
+
+    return OptimizeResult(res)
+
+
+def _unshuffle(array, n):
```

Comment:
It looks like this function is calculating the inverse permutation. This is simply
```
inv_perm = np.argsort(perm)
```
I would remove this function altogether and use `argsort` in the main body.